### PR TITLE
Adds a field processor which operates on dates

### PR DESCRIPTION
**Pull request by Craigacp**
### Description
Adds a `FieldProcessor` which parses a date value into a variety of features.

This also adds a new `GeneratedFeatureType` as date values produce integer features, along with adding locale generation to `DateExtractor` and `OffsetDateTimeExtractor`.

### Motivation
Tribuo needs a way of using dates as features as well as example metadata.

### CODE REVIEW

File: Data/src/main/java/org/tribuo/data/columnar/extractors/DateExtractor.java [hunk 1]
```diff
@@ -77,8 +102,14 @@ public DateExtractor(String fieldName, String metadataName, DateTimeFormatter fo
     @Override
     public void postConfig() {
         super.postConfig();
+        Locale locale;
+        if ((localeLanguage == null) && (localeCountry == null)) {
+            locale = Locale.getDefault(Locale.Category.FORMAT);
+        } else {
+            locale = new Locale(localeLanguage,localeCountry);
```
Comment (JackSullivan) [inline Data/src/main/java/org/tribuo/data/columnar/extractors/DateExtractor.java:109]: Won't we end up here if one of `localeLanguage` and `localeCountry` are null?
Maybe if either one is null we use the default and log a warning.

File: Data/src/main/java/org/tribuo/data/columnar/extractors/OffsetDateTimeExtractor.java [hunk 2]
```diff
@@ -65,8 +88,14 @@ public OffsetDateTimeExtractor(String fieldName, String metadataName, String dat
     @Override
     public void postConfig() {
         super.postConfig();
+        Locale locale;
+        if ((localeLanguage == null) && (localeCountry == null)) {
+            locale = Locale.getDefault(Locale.Category.FORMAT);
+        } else {
+            locale = new Locale(localeLanguage,localeCountry);
```
Comment (JackSullivan) [inline Data/src/main/java/org/tribuo/data/columnar/extractors/OffsetDateTimeExtractor.java:95]: Won't we end up here if one of `localeLanguage` and `localeCountry` are null?
Maybe if either one is null we use the default and log a warning.

**Review by JackSullivan — COMMENTED** [review]
Looks good apart from the minor comments inline on Locale nulls.

File: Data/src/main/java/org/tribuo/data/columnar/extractors/OffsetDateTimeExtractor.java (hunk 2 above)
Comment (Craigacp) [reply Data/src/main/java/org/tribuo/data/columnar/extractors/OffsetDateTimeExtractor.java:99]: I changed it so they throw `PropertyException`. The lack of default in locale is to ensure the default behaviour is the same in Tribuo 4.2+ as it was in 4.1 and 4.0.

File: Data/src/main/java/org/tribuo/data/columnar/extractors/DateExtractor.java (hunk 1 above)
Comment (Craigacp) [reply Data/src/main/java/org/tribuo/data/columnar/extractors/DateExtractor.java:114]: Changed to throw `PropertyException`.

**Review by JackSullivan — APPROVED** [review]
Looks good to me

**Pull request merged**